### PR TITLE
fix(legacy-rpc): eth_getLogs by blockHash returns empty array for local blocks with no matching events

### PR DESCRIPTION
**Pull request by JimmyShi22**
Previously, querying eth_getLogs with a blockHash filter would forward to legacy if the local result was empty. This caused "block not found" errors when the block existed locally but had no matching events, since legacy doesn't hold those blocks.

Fix: check block existence via eth_getBlockByHash before deciding routing.
- Block found locally → return local result directly (even if logs are [])
- Block not found locally → forward to legacy

Also adds unit tests for all three routing branches and an e2e regression test covering both the empty-logs case and a block with ERC20 Transfer events.

### CODE REVIEW

**Comment by JimmyShi22** [comment]
@claude review this

File: crates/legacy-rpc/src/get_logs.rs [hunk 1]
```diff
@@ -292,15 +292,19 @@ where
...
-                res
-            } else {
-                debug!(target:"xlayer_legacy_rpc", "method = eth_getLogs, forward to legacy (empty or error)");
-                service.forward_to_legacy(req).await
+        Some(GetLogsParams::BlockHash(block_hash)) => {
+            debug!(target:"xlayer_legacy_rpc", "method = eth_getLogs, checking if block exists locally...");
+            match service.call_eth_get_block_by_hash(&block_hash, false).await {
+                Ok(Some(_)) => {
+                    // Block exists locally; return local result even if logs are empty
+                    debug!(target:"xlayer_legacy_rpc", "method = eth_getLogs, block exists locally, using local result");
```
[the remark this answers is not in the document]
Comment (JimmyShi22) [reply crates/legacy-rpc/src/get_logs.rs:300]: fixed

**Comment by JimmyShi22** [comment]
@claude pls review again

File: crates/legacy-rpc/src/get_logs.rs [hunk 2]
```diff
@@ -292,15 +292,25 @@ where
...
-            let res = inner.call(req.clone()).await;
-            if res.is_success() && !is_result_empty(&res) {
-                debug!(target:"xlayer_legacy_rpc", "method = eth_getLogs, success response = {res:?}");
-                res
-            } else {
-                debug!(target:"xlayer_legacy_rpc", "method = eth_getLogs, forward to legacy (empty or error)");
-                service.forward_to_legacy(req).await
+        Some(GetLogsParams::BlockHash(block_hash)) => {
+            debug!(target:"xlayer_legacy_rpc", "method = eth_getLogs, checking if block exists locally...");
+            match service.call_eth_get_block_by_hash(&block_hash, false).await {
```
Comment (sieniven) [inline crates/legacy-rpc/src/get_logs.rs:297]: How about we pass the chainstate provider here directly instead of calling the eth_getBlockByHash handler instead?

Maybe a better way will be to pass an instance of the provider into the legacy service itself?

Comment (JimmyShi22) [reply crates/legacy-rpc/src/get_logs.rs:297]: This is consistent with how handle_block_param_methods already checks block existence — it calls call_eth_get_block_by_hash via inner in the same way (service.rs). We're reusing the same established pattern rather than introducing a new dependency.

Comment (JimmyShi22) [reply crates/legacy-rpc/src/get_logs.rs:297]: Injecting a chainstate provider would require adding a new field/generic to LegacyRpcRouterService and coupling it to reth internals. The gain is saving one in-process JSON round-trip, which has no network overhead and is negligible in practice. The current abstraction boundary — the service only needs inner: S — is worth keeping.

**Review by sieniven — APPROVED** [review]

**Review by sieniven — APPROVED** [review]

**Comment by louisliu2048** [comment]
@claude  review this pr

**Comment by JimmyShi22** [comment]
Test Result: SCCESS

```
$ bash ./scripts/test_block_hash_logs.sh
================================================================
 eth_getLogs Block Hash Race Condition Test
================================================================
 RPC URL     : http://localhost:8124
 Iterations  : 500
 Concurrency : 10
 Threshold   : 1% failure rate
 Timeout     : 5s per request
================================================================

  Progress: 500 / 500

================================================================
 RESULTS
================================================================
 Total iterations:              500
 Success:                       500
 Failures (total):              0
   - get_block failed:          0
   - get_logs error:            0
   - get_logs null:             0
   - timeout:                   0
 Failure rate:                  0.00%
 Duration:                      28s

================================================================
 RESULT: PASS  (failure rate 0.00% <= threshold 1%)
================================================================
```

Test script:
``` shell
#!/usr/bin/env bash
# test_block_hash_logs.sh
#
# Verify the eth_getLogs(blockHash) race condition fix.
#
# Problem: When calling eth_getBlockByNumber("latest") and then immediately querying
#          eth_getLogs with the returned blockHash, ~10% of requests fail because
#          the block is not yet indexed in the log index.
#
# Usage:
#   ./test_block_hash_logs.sh [OPTIONS]
#
# Options:
#   -u, --rpc-url URL        RPC endpoint URL (default: http://localhost:8124)
#   -n, --iterations N       Total number of test iterations (default: 500)
#   -c, --concurrency N      Number of concurrent workers (default: 10)
#   -t, --threshold PCT      Failure rate threshold; above this = FAIL (default: 1)
#   -T, --timeout SECS       Per-request timeout in seconds (default: 5)
#   -v, --verbose            Print details for each failure
#   -h, --help               Show this help message
#
# Examples:
#   ./test_block_hash_logs.sh --rpc-url http://localhost:8124
#   ./test_block_hash_logs.sh --rpc-url http://localhost:8124 -n 1000 -c 20 -v

set -uo pipefail

# ── Defaults ─────────────────────────────────────────────────────────────────
RPC_URL="http://localhost:8124"
ITERATIONS=500
CONCURRENCY=10
THRESHOLD=1
TIMEOUT=5
VERBOSE=false

# ── Argument parsing ──────────────────────────────────────────────────────────
usage() {
    sed -n '3,25p' "$0" | sed 's/^# \{0,1\}//'
    exit 0
}

while [[ $# -gt 0 ]]; do
    case "$1" in
        -u|--rpc-url)    RPC_URL="$2";     shift 2 ;;
        -n|--iterations) ITERATIONS="$2";  shift 2 ;;
        -c|--concurrency)CONCURRENCY="$2"; shift 2 ;;
        -t|--threshold)  THRESHOLD="$2";   shift 2 ;;
        -T|--timeout)    TIMEOUT="$2";     shift 2 ;;
        -v|--verbose)    VERBOSE=true;     shift   ;;
        -h|--help)       usage ;;
        *) echo "Unknown option: $1" >&2; exit 1 ;;
    esac
done

# ── Dependency check ──────────────────────────────────────────────────────────
for cmd in curl jq; do
    if ! command -v "$cmd" &>/dev/null; then
        echo "ERROR: '$cmd' is required but not installed." >&2
        exit 1
    fi
done

# ── Temp directory (cleaned up on exit) ──────────────────────────────────────
TMPDIR_WORK=$(mktemp -d)
trap 'rm -rf "$TMPDIR_WORK"' EXIT

# One file per result category; each worker appends a line on hit
SUCCESS_FILE="$TMPDIR_WORK/success"
FAIL_GET_BLOCK_FILE="$TMPDIR_WORK/fail_get_block"
FAIL_GET_LOGS_FILE="$TMPDIR_WORK/fail_get_logs"
FAIL_TIMEOUT_FILE="$TMPDIR_WORK/fail_timeout"
FAIL_NULL_FILE="$TMPDIR_WORK/fail_null"
DETAILS_FILE="$TMPDIR_WORK/details"

touch "$SUCCESS_FILE" "$FAIL_GET_BLOCK_FILE" "$FAIL_GET_LOGS_FILE" \
      "$FAIL_TIMEOUT_FILE" "$FAIL_NULL_FILE" "$DETAILS_FILE"

# ── Single iteration function ─────────────────────────────────────────────────
# Each iteration: eth_getBlockByNumber("latest") → eth_getLogs({blockHash}) immediately
run_iteration() {
    local id=$1

    # Step 1: fetch latest block
    local block_resp
    block_resp=$(curl -s -m "$TIMEOUT" -X POST \
        -H "Content-Type: application/json" \
        --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest",false],"id":1}' \
        "$RPC_URL" 2>/dev/null)

    local curl_rc=$?
    if [[ $curl_rc -ne 0 || -z "$block_resp" ]]; then
        echo 1 >> "$FAIL_TIMEOUT_FILE"
        [[ "$VERBOSE" == true ]] && echo "[#$id] FAIL(timeout) eth_getBlockByNumber" >> "$DETAILS_FILE"
        return
    fi

    # Check for RPC-level error in block response
    local rpc_err
    rpc_err=$(echo "$block_resp" | jq -r '.error.message // empty' 2>/dev/null)
    if [[ -n "$rpc_err" ]]; then
        echo 1 >> "$FAIL_GET_BLOCK_FILE"
        [[ "$VERBOSE" == true ]] && echo "[#$id] FAIL(get_block) $rpc_err" >> "$DETAILS_FILE"
        return
    fi

    # Extract blockHash
    local block_hash
    block_hash=$(echo "$block_resp" | jq -r '.result.hash // empty' 2>/dev/null)
    if [[ -z "$block_hash" || "$block_hash" == "null" ]]; then
        echo 1 >> "$FAIL_GET_BLOCK_FILE"
        [[ "$VERBOSE" == true ]] && echo "[#$id] FAIL(get_block) null hash" >> "$DETAILS_FILE"
        return
    fi

    # Step 2: fetch logs by blockHash — fired immediately after step 1
    local logs_resp
    logs_resp=$(curl -s -m "$TIMEOUT" -X POST \
        -H "Content-Type: application/json" \
        --data "{\"jsonrpc\":\"2.0\",\"method\":\"eth_getLogs\",\"params\":[{\"blockHash\":\"$block_hash\"}],\"id\":2}" \
        "$RPC_URL" 2>/dev/null)

    curl_rc=$?
    if [[ $curl_rc -ne 0 || -z "$logs_resp" ]]; then
        echo 1 >> "$FAIL_TIMEOUT_FILE"
        [[ "$VERBOSE" == true ]] && echo "[#$id] FAIL(timeout) eth_getLogs blockHash=$block_hash" >> "$DETAILS_FILE"
        return
    fi

    # Check for RPC-level error in logs response
    rpc_err=$(echo "$logs_resp" | jq -r '.error.message // empty' 2>/dev/null)
    if [[ -n "$rpc_err" ]]; then
        echo 1 >> "$FAIL_GET_LOGS_FILE"
        [[ "$VERBOSE" == true ]] && echo "[#$id] FAIL(get_logs) $rpc_err  blockHash=$block_hash" >> "$DETAILS_FILE"
        return
    fi

    # Verify result is an array (empty array is valid — block exists but has no logs)
    local result_type
    result_type=$(echo "$logs_resp" | jq -r '
        if .result == null then "null"
        elif (.result | type) == "array" then "array"
        else "other"
        end' 2>/dev/null)

    case "$result_type" in
        array)
            echo 1 >> "$SUCCESS_FILE"
            ;;
        null)
            echo 1 >> "$FAIL_NULL_FILE"
            [[ "$VERBOSE" == true ]] && echo "[#$id] FAIL(null_result) blockHash=$block_hash" >> "$DETAILS_FILE"
            ;;
        *)
            echo 1 >> "$FAIL_GET_LOGS_FILE"
            [[ "$VERBOSE" == true ]] && echo "[#$id] FAIL(unexpected) result=$result_type  blockHash=$block_hash" >> "$DETAILS_FILE"
            ;;
    esac
}

export -f run_iteration
export RPC_URL TIMEOUT VERBOSE
export SUCCESS_FILE FAIL_GET_BLOCK_FILE FAIL_GET_LOGS_FILE FAIL_TIMEOUT_FILE FAIL_NULL_FILE DETAILS_FILE

# ── Header ────────────────────────────────────────────────────────────────────
echo "================================================================"
echo " eth_getLogs Block Hash Race Condition Test"
echo "================================================================"
echo " RPC URL     : $RPC_URL"
echo " Iterations  : $ITERATIONS"
echo " Concurrency : $CONCURRENCY"
echo " Threshold   : ${THRESHOLD}% failure rate"
echo " Timeout     : ${TIMEOUT}s per request"
echo "================================================================"
echo ""

# ── Run test iterations in batches ───────────────────────────────────────────
START_TIME=$(date +%s)
batch_count=0

for i in $(seq 1 "$ITERATIONS"); do
    run_iteration "$i" &
    (( batch_count++ ))

    # Wait for the current batch to finish before launching the next one
    if (( batch_count >= CONCURRENCY )); then
        wait
        batch_count=0

        done_count=$(( $(wc -l < "$SUCCESS_FILE") + $(wc -l < "$FAIL_GET_BLOCK_FILE") + \
                       $(wc -l < "$FAIL_GET_LOGS_FILE") + $(wc -l < "$FAIL_TIMEOUT_FILE") + \
                       $(wc -l < "$FAIL_NULL_FILE") ))
        printf "\r  Progress: %d / %d" "$done_count" "$ITERATIONS"
    fi
done

# Wait for remaining background jobs
wait
printf "\r  Progress: %d / %d\n\n" "$ITERATIONS" "$ITERATIONS"

END_TIME=$(date +%s)
DURATION=$(( END_TIME - START_TIME ))

# ── Aggregate results ─────────────────────────────────────────────────────────
count_lines() { wc -l < "$1" | tr -d ' '; }

N_SUCCESS=$(count_lines "$SUCCESS_FILE")
N_FAIL_BLOCK=$(count_lines "$FAIL_GET_BLOCK_FILE")
N_FAIL_LOGS=$(count_lines "$FAIL_GET_LOGS_FILE")
N_FAIL_TIMEOUT=$(count_lines "$FAIL_TIMEOUT_FILE")
N_FAIL_NULL=$(count_lines "$FAIL_NULL_FILE")
N_TOTAL=$(( N_SUCCESS + N_FAIL_BLOCK + N_FAIL_LOGS + N_FAIL_TIMEOUT + N_FAIL_NULL ))
N_FAIL=$(( N_FAIL_BLOCK + N_FAIL_LOGS + N_FAIL_TIMEOUT + N_FAIL_NULL ))

# Floating-point failure rate via awk
if (( N_TOTAL > 0 )); then
    FAIL_RATE=$(awk "BEGIN { printf \"%.2f\", $N_FAIL / $N_TOTAL * 100 }")
else
    FAIL_RATE="0.00"
fi

# ── Print report ──────────────────────────────────────────────────────────────
echo "================================================================"
echo " RESULTS"
echo "================================================================"
printf " %-30s %d\n"  "Total iterations:"      "$N_TOTAL"
printf " %-30s %d\n"  "Success:"               "$N_SUCCESS"
printf " %-30s %d\n"  "Failures (total):"      "$N_FAIL"
printf " %-30s %d\n"  "  - get_block failed:"  "$N_FAIL_BLOCK"
printf " %-30s %d\n"  "  - get_logs error:"    "$N_FAIL_LOGS"
printf " %-30s %d\n"  "  - get_logs null:"     "$N_FAIL_NULL"
printf " %-30s %d\n"  "  - timeout:"           "$N_FAIL_TIMEOUT"
printf " %-30s %s%%\n" "Failure rate:"         "$FAIL_RATE"
printf " %-30s %ds\n" "Duration:"              "$DURATION"
echo ""

# Print per-failure details in verbose mode
if [[ "$VERBOSE" == true && -s "$DETAILS_FILE" ]]; then
    echo "----------------------------------------------------------------"
    echo " Failure Details"
    echo "----------------------------------------------------------------"
    cat "$DETAILS_FILE"
    echo ""
fi

# ── Pass / Fail verdict ───────────────────────────────────────────────────────
echo "================================================================"
PASS=$(awk "BEGIN { print ($FAIL_RATE <= $THRESHOLD) ? 1 : 0 }")
if (( N_TOTAL == 0 )); then
    echo " RESULT: INCONCLUSIVE — no iterations completed"
    exit 2
elif [[ "$PASS" == "1" ]]; then
    echo " RESULT: PASS  (failure rate ${FAIL_RATE}% <= threshold ${THRESHOLD}%)"
    echo "================================================================"
    exit 0
else
    echo " RESULT: FAIL  (failure rate ${FAIL_RATE}% > threshold ${THRESHOLD}%)"
    echo "================================================================"
    exit 1
fi

```

**Comment by JimmyShi22** [comment]
@claude review this pr even merged
